content/developer/addons/events-and-handlers.md

```diff
@@ -109,3 +106,188 @@ With this addon enabled, going to the URL `/discussions/kaboom` would now output
 
 Magic methods only work in classes that extend `Gdn_Pluggable`. For example, notice the `Gdn_Form` class does, but the `Gdn_Format` class does not. All models and controllers do.
+
+## Example Events
+
+### Inject the the current user's roles into every page
+
+Sometimes you may want to adjust parts of the template based on the roles the current user. This will inject gather the roles of the current user and inject them into the smarty template.
+
+```php
+public function base_render_before($sender) {
+    if (inSection('Dashboard')) {
+        return;
+    }
+
+    if(!val('UserRoles', $sender->Data)) {
+        $user = val('User', Gdn::controller());
+        if (!$user && Gdn::session()->isValid()) {
+            $user = Gdn::session()->User;
+        }
+        $userRoles = [];
+        if ($user) {
+            $userID = val('UserID', $user);
+            $userRolesData = Gdn::userModel()->getRoles($userID);
+            if ($userRolesData->numRows() > 0) {
+                $userRoles = array_column($userRolesData->resultArray(), 'Name');
+            }
+        }
+        $sender->setData('UserRoles', $userRoles);
+    }
+}
+```
+
+### Inject a conditional based on roles
+
+Sometimes you may not need all the roles in the page. Let's say you wanted to make the page appear differently for a user with a certain role. Instead of injection all of the roles into the template and doing the conditional there, you can do it in the themehooks and inject just the boolean value you need.
+
+
+```php
+public function base_render_before($sender) {
+    if (inSection('Dashboard')) {
+        return;
+    }
+    
+    $roles = val('UserRoles', $sender->Data));
+    if (!$roles) {
+        $user = val('User', Gdn::controller());
+        if (!$user && Gdn::session()->isValid()) {
+            $user = Gdn::session()->User;
+        }
+        $userID = val('UserID', $user);
+        $roles = Gdn::userModel()->getRoles($userID)->resultArray();
+    }
+
+    $roleNames = array_column($roles, 'Name');
+    $isSuperSpecialRole = in_array("SuperSpecialRole", $roleNames);
+    $sender->setData("isSuperSpecialRole", $isSuperSpecialRole);
+}
+```
+
+### Create an additional settings page
+
+This example creates a custom dashboard page that can set a few configuration options. You would then need to use these set configuration values in other hooks to customize your site.
+
+```php
+class MySitePlugin extends Gdn_Plugin {
+     
+    /**
+     * Create the `/settings/example` page to host our custom settings.
+     *
+     * @param SettingsController $sender
+     */
+    public function settingsController_example_create($sender) {
+        $sender->permission('Garden.Settings.Manage');
+
+        $configurationModule = new ConfigurationModule($sender);
+        $configurationModule->initialize([
+            'ExmapleSite.BackgroundColour' => ['Control' => 'TextBox', 'Options' => ['class' => 'InputBox BigInput']],
```

Comment:
Do we have documentation anywhere on the types of inputs we accept? If we don't, we should make one. We should add a note with a link to that page.

Reply:
We don't currently have proper documentation for Gdn_Form or the Configuration Module. These documents are a bit of a project in of themselves, so for now I'm going to link to the Gdn_Form implementation and some additional examples in vanilla where we use the ConfigurationModule.

I'll open an issue for the Gdn_Form docs though.

Reply:
Additionally I've added a list of the support form components in the ConfigurationModule